### ISSUE TITLE
Query string includes all available filters causing an error on refresh

### ISSUE THREAD

**Issue by archilex** [laravel-filament/filament#7129]
### Package

filament/filament

### Package Version

v3

### Laravel Version

v9.19

### Livewire Version

v3

### PHP Version

PHP 8.1

### Problem description

In v3 when you select a filter, the query string will then populate with **all** the available filters, setting the others to `null`. While this pollutes the query string, the bigger issue is that with certain types of filters if you then refresh the page you will get an error as `null` is not an acceptable value.

In v2, only the active filter would be present in the query string.

### Expected behavior

The query string should only include active filters

### Steps to reproduce

1. On the users page of the demo app filter by just one of the two dates, leaving the other date field alone.
2. The query string will populate with the unfiltered date setting it to `null`: `http://demov3.test/users?tableFilters[created_at][created_from]=2023-07-01&tableFilters[created_at][created_until]=null`
3. Refresh the page
4. Error: `Could not parse 'null': Failed to parse time string (null) at position 0 (n): The timezone could not be found in the database`

### Reproduction repository

https://github.com/archilex/demo

**Comment by archilex** [comment]
You can pull down the repo, but you can also see this on the filament v3 demo as well.

**Comment by zepfietje** [comment]
> You can pull down the repo, but you can also see this on the filament v3 demo as well.

Thanks, I've confirmed this on the demo indeed.

> In v2, only the active filter would be present in the query string.

I'm pretty sure this is caused by the change in how Livewire handles query parameters by default in v3.

**Comment by archilex** [comment]
I agree. I've tried some work arounds but it seems to be a livewire issue. 

Probably need to add a failing test so they can work on it. I would, but I'm slammed for the next week.

**Comment by zepfietje** [comment]
Is it really a Livewire bug instead of designed behavior though?

**Comment by ariaieboy** [comment]
> Is it really a Livewire bug instead of designed behavior though?

where can we access to the V3 demo?

**Comment by zepfietje** [comment]
> where can we access to the V3 demo?

Pull it in locally and use the 3.x branch.

**Comment by archilex** [comment]
I don't think this is by design as it seems like LW3's new intention is to NOT show query string parameters since they got rid of `exclude` and now you have to use `keep` to have them appear in the url.

**Comment by ariaieboy** [comment]
It's a bug in the livewire JS part if you set the history parameter to true it will not change the empty input to null. 
```php
class Test extends Component
{
    #[Url(history: true,keep: false)]
    public string $search;

    public function render()
    {
        return <<<'HTML'
        <div>
            <input type="text" wire:model.live="search">
        </div>
        HTML;
    }
}
```
if you hit the URL pointing to this component like `localhost/test?search=` it will not change or remove the query parameter. 
but if you change the `history` to false and hit the URL it will change the URL to `localhost/test.search=null` 
and it's two separate bugs. first it should not set value null on empty input and it must remove the empty string from query params since we didn't pass the keep option as true.

**Comment by zepfietje** [comment]
So are we sure this is 100% a Livewire bug?
And if so, is there a workaround or not?

**Comment by ariaieboy** [comment]
> So are we sure this is 100% a Livewire bug? And if so, is there a workaround or not?

if you set the history to true it will not change the value to `null` so the problem will fixs for now but it's a bug that must be fixed by livewire. 
And it's not related to livewire backend and I am not that good with JS stuff. 

I've created a bug report on livewire repo https://github.com/livewire/livewire/discussions/5932

**Comment by zepfietje** [comment]
@danharrin, should we just wait for this to be fixed in Livewire?

**Comment by ariaieboy** [comment]
This bug get fixed by livewire 
https://github.com/livewire/livewire/commit/b6bdebad0c90c2e6486367e52d538f5490b2ad7d

**Comment by archilex** [comment]
Thanks @ariaieboy for your work on the LW issue and failing test to get this fixed!

**Comment by archilex** [comment]
So unfortunately, while this is fixed in Livewire, this is still an issue in Filament with the tableFilters array. You can test this the same way by opening the v3 demo and setting just one of the filters. All the other filters are still set to null. 

This time I don't believe its a livewire issue. Previously if you were to go set the querystring to just `tableFilters[trashed][value]=` livewire would change it to `tableFilters[trashed][value]=null` and add all the other filters as well. With the livewire fix, it correctly removes it from the url. However, if you set just one of the filters, then all the other filters are still added to the query string with null values.

**Comment by sumardi** [comment]
https://github.com/filamentphp/filament/assets/315603/b4a47198-174b-49b4-8891-75abffb68af4

I think with this new Livewire, default state need to be set as empty string `""`, not `null`.

**Comment by archilex** [comment]
EDIT: Ignore this. See next comment


I can now confirm that has something to do with Filament. I finally had some time to pull down the latest livewire v3 and run some browser tests there. I had thought that maybe it had something to do with the way livewire parsed associative arrays, but the browers tests I created in livewire show that LW *is* removing any `null` value from the query string, even if it's within an associative array. LW is also *not* adding the other key/value null pairs of the associative array which *is* happending in Filament.

I'll keep digging into Filament to see.

**Comment by archilex** [comment]
Ok, for real this time, I can confirm that this is indeed a livewire bug. I've filed a failing test on livewire. My previous browser test wasn't testing the actual problem, but I've worked out another test that consistently shows the issue.

**Comment by KennedyTedesco** [comment]
@archilex, Thank you for addressing this issue. It's a significant bug for my application, making all tables filters almost unusable. I hope we can find a solution soon. ❤️

**Comment by KennedyTedesco** [comment]
All my tables extend a `BaseTableComponent`. So, until we find a solution on Livewire, I've mitigated the issue by doing this:

```php
public function boot(): void
{
    $this->replaceNullValues($this->tableFilters)
}

protected function replaceNullValues(array &$data): void
{
    foreach ($data as &$value) {
        if (is_array($value)) {
            $this->replaceNullValues($value);
        } elseif ($value === "null") {
            $value = null;
        }
    }
}
```

I know this isn't beautiful, but it works for now. Haha!

**Comment by jvitasek** [comment]
@KennedyTedesco 
This didn't work for me exactly, had to add a check for `tableFilters` (otherwise got errors), but thanks for the help! Also I didn't know where to put this code right away.

In the list resource page, e.g. `OrderResource/Pages/ListOrders.php`:
```php
public function boot(): void
{
    if ($this->tableFilters) {
        $this->replaceNullValues($this->tableFilters);
    }
}

protected function replaceNullValues(array &$data): void
{
    foreach ($data as &$value) {
        if (is_array($value)) {
            $this->replaceNullValues($value);
        } elseif ($value === "null") {
            $value = null;
        }
    }
}
```

**Comment by sakanjo** [comment]
I've tested it on my end and it doesn't error on refresh, but the query strings are all included.

versions:
`livewire/livewire: v3.0.0-beta.9`
`filament/filament: 3.x-dev`

**Comment by KennedyTedesco** [comment]
@jvitasek Oh, you're right. It's because I use Filament components (tables, forms, etc) in a standalone mode inside Livewire components, so, in my case, `tableFilters` are always defined. In your case, I don't know where you should put this, because I have never used Filament full admin mode. I hope someone here can guide you.

**Comment by jvitasek** [comment]
@KennedyTedesco Oh, I didn't know about the standalone mode. The code works OK (for Filament full admin) as I posted it and fixed the issue. Thanks again for your snippet!

**Comment by archilex** [comment]
> I've tested it on my end and it doesn't error on refresh, but the query strings are all included.

Yes, the error with null values in dates has been fixed in Filament. Now we're just waiting on LW to fix the null values appearing in the query string.

**Comment by jvitasek** [comment]
So I also encountered problems when dealing with Toggle/Checkbox filters. Livewire doesn't convert "true"/"false" to bool values, so the filters don't work. Here's the updated code:
```php
public function boot(): void
{
    if ($this->tableFilters) {
        $this->replaceNullValues($this->tableFilters);
    }
}

protected function replaceNullValues(array &$data): void
{
    foreach ($data as &$value) {
        if (is_array($value)) {
            $this->replaceNullValues($value);
        } elseif ($value === 'null') {
            $value = null;
        } elseif ($value === 'false') {
            $value = false;
        } elseif ($value === 'true') {
            $value = true;
        }
    }
}
```

**Comment by zepfietje** [comment]
@archilex, if a fix for your failing test (https://github.com/livewire/livewire/pull/6188) gets merged into Livewire, this issue should be fixed?

**Comment by archilex** [comment]
@zepfietje I sure hope so!

**Comment by archilex** [comment]
In light of Caleb's comment that he won't be able to address this in Livewire in the near future, should I look at implementing a temporary fix in Filament until it's fixed in Livewire?

Even though the dateFilter is no longer throwing errors, this is still a problem with SelectFilter, TrashedFilter, TernaryFilter, and Filter with forms with textFields on refresh.

**Comment by bernhardh** [comment]
Yes please. I am running into the same issue :(

**Comment by monointeractive** [comment]
My workaround 

app.js:
```
import endsWith from 'lodash/endsWith';
window.history.pushState = ((pushState)=>{
    let removeEmptys = (fullurl) =>{
        let [url,rest=''] = fullurl.split('?');
        let [search,hash=''] = rest.split('#');
        let params = search.split('&');
        for(let i in params){
            let check = String(params[i]).trim().toLowerCase();
            if(endsWith(check,'=null')) params[i] = null;
            if(endsWith(check,'=')) params[i] = null;
            if(endsWith(check,'[isactive]=false')) params[i] = null;
        }
        search = params.filter(Boolean).join('&').trim();

        fullurl = url;
        if(search) fullurl = `${fullurl}?${search}`;
        if(hash) fullurl = `${fullurl}#${hash}`;
        return fullurl;
    }

    return function(){
        const args = [...arguments];
        let url = (args?.[2]) || '';
        if(url?.indexOf?.('/') > -1){
            console.log(url);
            url = removeEmptys(url);
            args[2] = url;
        }
        pushState.apply(window.history,args);
    }
})(window.history.pushState);

```
